src/v/io/persistence.cc

```diff
@@ -0,0 +1,327 @@
+    auto ret = files_.try_emplace(
+      path, seastar::make_shared<memory_file>(this));
+    if (!ret.second) {
+        return seastar::make_exception_future<
+          seastar::shared_ptr<persistence::file>>(
+          std::filesystem::filesystem_error(
+            "File exists",
+            path,
+            std::error_code(EEXIST, std::system_category())));
+    }
+    return seastar::make_ready_future<seastar::shared_ptr<persistence::file>>(
+      ret.first->second);
+}
+
+seastar::future<seastar::shared_ptr<persistence::file>>
+memory_persistence::open(std::filesystem::path path) noexcept {
+    auto it = files_.find(path);
+    if (it == files_.end()) {
+        return seastar::make_exception_future<
+          seastar::shared_ptr<persistence::file>>(
+          std::filesystem::filesystem_error(
+            "File does not exist",
+            path,
+            std::error_code(ENOENT, std::system_category())));
+    }
+    return seastar::make_ready_future<seastar::shared_ptr<persistence::file>>(
+      it->second);
+}
+
+memory_persistence::memory_file::memory_file(memory_persistence* persistence)
+  : persistence_(persistence) {}
+
+seastar::future<size_t> memory_persistence::memory_file::dma_read(
+  uint64_t pos, char* buf, size_t len) noexcept {
+    if (auto err = check_alignment(
+          "dma_read",
+          pos,
+          buf,
+          len,
+          memory_dma_alignment(),
+          disk_read_dma_alignment());
+        err.has_value()) {
+        return std::move(err.value());
+    }
+    return read(pos, buf, len);
+}
+
+seastar::future<size_t> memory_persistence::memory_file::dma_write(
+  uint64_t pos, const char* buf, size_t len) noexcept {
+    if (auto err = check_alignment(
+          "dma_write",
+          pos,
+          buf,
+          len,
+          memory_dma_alignment(),
+          disk_write_dma_alignment());
+        err.has_value()) {
+        return std::move(err.value());
+    }
+    return write(pos, buf, len).then([this, pos](size_t written) {
+        size_ = std::max(size_, pos + written);
+        return written;
+    });
+}
+
+seastar::future<> memory_persistence::memory_file::close() noexcept {
+    co_return;
+}
+
+uint64_t
+memory_persistence::memory_file::disk_read_dma_alignment() const noexcept {
+    return persistence_->disk_read_dma_alignment_;
+}
+
+uint64_t
+memory_persistence::memory_file::disk_write_dma_alignment() const noexcept {
+    return persistence_->disk_write_dma_alignment_;
+}
+
+uint64_t
+memory_persistence::memory_file::memory_dma_alignment() const noexcept {
+    return persistence_->memory_dma_alignment_;
+}
+
+seastar::future<>
+memory_persistence::memory_file::ensure_capacity(size_t size) {
+    // futurized to avoid reactor stalls for large capacity expansions
+    return seastar::do_until(
+      [this, size] { return capacity_ >= size; },
+      [this] {
+          data_.emplace(capacity_, chunk_size);
+          capacity_ += chunk_size;
```

Comment:
question: Should the chunk size be related to alignment at all?

Reply:
it shouldn't matter. chunk size is all about unit of capacity. data ultimately gets memcpy'd in/out so actual address of backing store doesn't matter.